File: src/gui/display.rs
#![allow(dead_code)]

extern crate pancurses;

use super::{pad::Pad, scrollable::Scrollable};
use pancurses::*;

#[derive(Debug)]
pub struct Display {
	parent: Option<Box<Display>>,
	pad: Pad,
	data: Vec<String>,
	title: String,
	xoffset: i32,
	yoffset: i32,
	xsize: i32,
	ysize: i32,

	// scrolling functionality
	index: i32,
}

impl Display {
	/// Creates a new [`Display`] object, this acts as the root [`Display`].
	pub fn new(win: &Window, title: String, data: Vec<String>) -> Self {
		// getting max x& y
		let (y, x) = win.get_max_yx();

		// creating the pad object
		let pad: Pad = match Pad::new(std::cmp::max(y, data.len() as i32), x / 3) {
			Ok(p) => p,
			Err(e) => {
				endwin();
				panic!("There was an error creating a pad\n{:?}", e);
			}
		};

		// creating struct
		let d = Display {
			parent: None,
			pad,
			data,
			title,
			xoffset: 0,
			yoffset: 0,
			xsize: x / 3,
			ysize: y,
			index: 0,
		};

		// drawing the contents of `Data`
		d.setup();

		// returning
		d
	}

	/// This function draws the contents of `Display.data` onto the pad
	fn setup(&self) {
		for i in 0..(self.data.len() as i32) {
			// borrowing the string from the array
			let s: &String = &self.data[i as usize];
			let l = s.len().try_into().unwrap_or(0);

			// writing the slice
			self.pad
				.mvaddnstr(self.yoffset + i, self.xoffset, s, l)
				.expect(&format!(
					"Error in display.setup() calling pad.mvaddnstr({}, {}, {}, {})",
					self.yoffset + i,
					self.xoffset,
					s,
					l
				));
		}

		// highlighting the first line
		self.pad.set_highlight(0).expect({
			endwin();
			"Error in display.setup(), highlighting line 0 of pad"
		});
	}

    /// Makes a call to the internal [`Pad`] to write a `n` bytes of a string to an `(y, x)` location.
    pub fn mvaddnstr(&self, y: i32, x: i32, s: &String, n: i32) -> Result<(), i32> {
        self.pad.mvaddnstr(y, x, s, n)
    }

	/// draws this [`Display`] to the screen.
	pub fn refresh(&self) {
		let scroll_mod = match self.index > (self.ysize - 1) {
			true => std::cmp::min(
				self.index - self.ysize + 1,
				self.data.len() as i32 - self.ysize,
			),
			false => 0,
		};

		self.pad
			.refresh(scroll_mod, 0, 0, 0, self.ysize - 1, self.xsize)
			.expect(&format!(
				"Error in display.refresh() calling pad.refresh({}, {}, {}, {}, {}, {})",
				scroll_mod,
				0,
				0,
				0,
				self.ysize - 1,
				self.xsize
			))
	}

	/// Calls the internal [`Pad`]'s getch.
	pub fn getch(&self) -> Option<Input> {
		self.pad
			.getch()
			.expect("Error in display.getch() calling pad.getch() = {}")
	}

	pub fn keypad(&self, state: bool) {
		self.pad
			.keypad(state)
			.expect("Error in display.keypad() calling pad.keypad()");
	}

    /// Makes a call to the internal `Pad` object which "moves" the cursor up one line (within
    /// the boundaries of the window.
	pub fn scroll_up(&mut self) {
		let new = std::cmp::max(self.index - 1, 0);

		// setting the old index to normal
		self.pad.remove_highlight(self.index).expect({
			endwin();
			&format!(
				"Error in display.scroll_up(), highlighting line {} of pad",
				self.index
			)
		});

		// setting the new index to normal
		self.pad.set_highlight(new).expect({
			endwin();
			&format!(
				"Error in display.scroll_up(), highlighting line {} of pad",
				new
			)
		});

		// storing the new index
		self.index = new;
	}

    /// Makes a call to the internal `Pad` object which "moves" the cursor down one line (within
    /// the boundaries of the window.
	pub fn scroll_down(&mut self) {
		let new = std::cmp::min(self.index + 1, self.pad.dimensions().0 - 1);

		// setting the old index to normal
		self.pad.remove_highlight(self.index).expect({
			endwin();
			&format!(
				"Error in display.scroll_down(), highlighting line {} of pad",
				self.index
			)
		});

		// setting the new index to normal
		self.pad.set_highlight(new).expect({
			endwin();
			&format!(
				"Error in display.scroll_down(), highlighting line {} of pad",
				new
			)
		});

		// storing the new index
		self.index = new;
	}
}

//// Trait implementations for [`Display`]

impl Drop for Display {
	fn drop(&mut self) {
		return match& self.parent {
			Some(win) => {
				drop(win);
			}
			None => {}
		};
	}
}

File: src/gui/display.py
"""
the module for the Display object, which is a better wrapper around the ncurses
pad object
"""

# global imports
import curses

# local imports
from . import keys
from core.data import junk_data


class Display:
    def __init__(self, window, title, data, xoffset=0, parent=None):
        """
        creates a pad object based off of the window object
        """

        # from arguments
        self.window = window
        self.title = title
        self.data = data

        # getting the dimensions of the parent window (usually from stdscr)
        (ymax, xmax) = window.getmaxyx()

        # internal vars
        # setting the offset
        self.ysize = ymax
        self.xsize = int(xmax / 3)
        self.yoffset = 1
        self.xoffset = xoffset

        # scrolling functionality
        self.index = 0

        # setting a reference to the parent window object
        self.parent = parent

        # creating internal objects
        self.pad = curses.newpad(max(ymax, len(data)), self.xsize)

        # setting up the pad
        self.__setup__()

    def __setup__(self):
        """
        this function writes the contents of `self.data` to the `self.pad`
        object
        """

        # writing the title atop the pad's render box
        self.window.addnstr(0, self.xoffset, self.title, self.xsize)

        # writing self.data to the pad
        for i, s in enumerate(self.data):
            self.pad.addnstr(i, 0, s, self.xsize)

        # highlighting the first line
        self.set_highlight(0)

    def set_highlight(self, line):
        """
        sets a highlight at a given index
        """

        return self.pad.chgat(line, 0, curses.A_REVERSE)

    def remove_highlight(self, line):
        """
        removes the highlight at a given index
        """

        return self.pad.chgat(line, 0, curses.A_NORMAL)

    def spawn_child(self):
        """
        this function creates a new Display as a child from the current
        index of the current window
        """

        # creating new display
        d = Display(self.window, "Child", junk_data(150, 30),
                    xoffset=self.xsize, parent=self)

        # returning the new display
        return d

    def destroy(self):
        """
        destroying a window via. clearing it, refreshing, then deleting it
        """

        # remove the title
        self.window.addnstr(0, self.xoffset, " " * self.xsize, self.xsize)

        self.pad.clear()  # clear the pad
        self.refresh()  # refresh the Display object
        del self.pad  # delete the pad
        self.window.refresh()  # refresh the main window

    def handle_input(self, char):
        """
        this function makes an action based off of the provided input
        """

        match char:

            # scrolling keys
            case keys.ARROWDOWN:
                (ypad, _) = self.pad.getmaxyx()
                new = min(self.index + 1, ypad - 1)
                self.remove_highlight(self.index)
                self.set_highlight(new)
                self.index = new

            case keys.ARROWUP:
                new = max(self.index - 1, 0)
                self.remove_highlight(self.index)
                self.set_highlight(new)
                self.index = new

            case keys.PGDOWN:
                (ypad, _) = self.pad.getmaxyx()
                new = min(self.index + self.ysize, ypad - 1)
                self.remove_highlight(self.index)
                self.set_highlight(new)
                self.index = new

            case keys.PGUP:
                new = max(self.index - self.ysize, 0)
                self.remove_highlight(self.index)
                self.set_highlight(new)
                self.index = new

            case _:
                pass

    def refresh(self):
        """
        refreshes and redraws the pad to the screen
        """

        # calculating the offset from the top
        scroll_mod = 0
        if self.index > (self.ysize - 1 - self.yoffset):
            scroll_mod = min(self.index - self.ysize + 1,
                             len(self.data) - self.ysize) + self.yoffset

        # refreshing the pad
        self.pad.refresh(scroll_mod, 0,
                         self.yoffset, self.xoffset,
                         self.ysize - 1, self.xsize + self.xoffset)

